File: Fast_LISA_Waveforms/python/tdi2_response.py
"""Directed-link TDI-2 Michelson response for complex polarization carriers.

The 16 signed link terms per channel follow the SGS/PyTDI X2 convention.
The default evaluates six directed links at each output epoch and expands
retarded delay words to first order, matching the fast C response.  A source
callback receives sorted barycentric source times and returns analytic
``(hplus, hcross)`` arrays, each either ``(samples,)`` or
``(families, samples)``.  Outputs have shape ``(3, families, samples)``.
"""

# Copyright (C) 2026 Neil Cornish.
# SPDX-License-Identifier: GPL-3.0-or-later

from __future__ import annotations

import math
from collections.abc import Callable, Sequence

import numpy as np


_X_TERMS = (
    (1, 3,  1, ()),
    (3, 1,  1, (13,)),
    (1, 2,  1, (13, 31)),
    (2, 1,  1, (13, 31, 12)),
    (1, 2, -1, ()),
    (2, 1, -1, (12,)),
    (1, 3, -1, (12, 21)),
    (3, 1, -1, (12, 21, 13)),
    (1, 2,  1, (13, 31, 12, 21)),
    (2, 1,  1, (13, 31, 12, 21, 12)),
    (1, 3,  1, (13, 31, 12, 21, 12, 21)),
    (3, 1,  1, (13, 31, 12, 21, 12, 21, 13)),
    (1, 3, -1, (12, 21, 13, 31)),
    (3, 1, -1, (12, 21, 13, 31, 13)),
    (1, 2, -1, (12, 21, 13, 31, 13, 31)),
    (2, 1, -1, (12, 21, 13, 31, 13, 31, 12)),
)


def _position(splines: Sequence, times: np.ndarray, derivative: bool = False) -> np.ndarray:
    values = [(s.derivative()(times) if derivative else s(times)) for s in splines]
    return np.stack(values, axis=-1).reshape(len(times), 3, 3)


def _basis(latitude: float, longitude: float) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    costh = math.sin(latitude)
    sinth = math.sqrt(max(0.0, 1.0-costh*costh))
    cosph, sinph = math.cos(longitude), math.sin(longitude)
    u = np.array((-costh*cosph, -costh*sinph, sinth))
    v = np.array((sinph, -cosph, 0.0))
    propagation = np.array((-sinth*cosph, -sinth*sinph, -costh))
    return propagation, np.outer(v, v)-np.outer(u, u), np.outer(u, v)+np.outer(v, u)


def _reference_links(times: np.ndarray, splines: Sequence, propagation: np.ndarray,
                     eplus: np.ndarray, ecross: np.ndarray) -> dict[tuple[int, int], tuple[np.ndarray, ...]]:
    position = _position(splines, times)
    velocity = _position(splines, times, derivative=True)
    links = {}
    for receiver in range(3):
        for emitter in range(3):
            if receiver == emitter:
                continue
            radius = position[:, receiver]-position[:, emitter]
            r2 = np.einsum("ij,ij->i", radius, radius)
            rv = np.einsum("ij,ij->i", radius, velocity[:, emitter])
            v2 = np.einsum("ij,ij->i", velocity[:, emitter], velocity[:, emitter])
            light_time = (rv+np.sqrt(rv*rv+(1.0-v2)*r2))/(1.0-v2)
            retarded_time = times-light_time
            if (retarded_time[0] < splines[0].x[0] or
                    retarded_time[-1] > splines[0].x[-1]):
                raise ValueError("TDI-2 retarded link extends beyond the constellation grid")
            emitter_position = _position(splines, retarded_time)[:, emitter]
            emitter_velocity = _position(splines, retarded_time, derivative=True)[:, emitter]
            direction = position[:, receiver]-emitter_position
            direction /= np.linalg.norm(direction, axis=1)[:, None]
            numerator = np.einsum("ij,ij->i", direction,
                                  velocity[:, receiver]-emitter_velocity)
            denominator = 1.0-np.einsum("ij,ij->i", direction, emitter_velocity)
            dot_light_time = numerator/denominator
            kdot = direction @ propagation
            plus = -0.5*np.einsum("ni,nj,ij->n", direction, direction, eplus)/(1.0-kdot)
            cross = -0.5*np.einsum("ni,nj,ij->n", direction, direction, ecross)/(1.0-kdot)
            receiver_source = times-position[:, receiver] @ propagation
            emitter_source = retarded_time-emitter_position @ propagation
            receiver_rate = 1.0-velocity[:, receiver] @ propagation
            emitter_rate = (1.0-dot_light_time)*(1.0-emitter_velocity @ propagation)
            links[receiver, emitter] = (light_time, dot_light_time,
                                        receiver_source, emitter_source,
                                        receiver_rate, emitter_rate, plus, cross)
    return links


def monochromatic_tdi2_transfer(
    times: np.ndarray,
    pixel_frequency: np.ndarray,
    pixel_time_columns: np.ndarray,
    latitude: float,
    longitude: float,
    position_splines: Sequence,
    *,
    maximum_derivative: int = 0,
) -> np.ndarray:
    """Return plus/cross X2/Y2/Z2 symbols and their frequency derivatives.

    The output shape is ``(maximum_derivative+1, 3, 2, pixels)``. Each symbol
    multiplies a positive-phase analytic polarization at the pixel output
    time. Geometry and directed links are shared across all frequency pixels
    at a given time, while delay phasors are evaluated at the pixel frequency.
    """
    times = np.asarray(times, dtype=np.float64)
    frequency = np.asarray(pixel_frequency, dtype=np.float64)
    columns = np.asarray(pixel_time_columns, dtype=np.int64)
    if maximum_derivative not in (0, 1, 2):
        raise ValueError("maximum_derivative must be 0, 1, or 2")
    if times.ndim != 1 or np.any(np.diff(times) <= 0.0):
        raise ValueError("times must be a strictly increasing vector")
    if frequency.ndim != 1 or columns.shape != frequency.shape or \
       np.any(columns < 0) or np.any(columns >= times.size):
        raise ValueError("pixel frequencies and time columns must align")
    propagation, eplus, ecross = _basis(latitude, longitude)
    links = _reference_links(times, position_splines, propagation, eplus, ecross)
    local_time = times[columns]
    response = np.zeros((maximum_derivative+1, 3, 2, frequency.size),
                        dtype=np.complex128)
    for channel in range(3):
        for receiver_label, emitter_label, sign, delay_words in _X_TERMS:
            lag = np.zeros(times.size)
            jacobian = np.ones(times.size)
            for label in delay_words:
                receiver = (label//10-1+channel) % 3
                emitter = (label%10-1+channel) % 3
                light_time, derivative = links[receiver, emitter][:2]
                lag += light_time-derivative*lag
                jacobian *= 1.0-derivative
            receiver = (receiver_label-1+channel) % 3
            emitter = (emitter_label-1+channel) % 3
            link = links[receiver, emitter]
            receiver_delay = (link[2]-link[4]*lag)[columns]-local_time
            emitter_delay = (link[3]-link[5]*lag)[columns]-local_time
            receiver_phase = np.exp(2j*math.pi*frequency*receiver_delay)
            emitter_phase = np.exp(2j*math.pi*frequency*emitter_delay)
            weight = sign*jacobian[columns]
            for order in range(maximum_derivative+1):
                factor = (2j*math.pi)**order
                difference = factor*(receiver_delay**order*receiver_phase-
                                     emitter_delay**order*emitter_phase)
                response[order, channel, 0] += weight*link[6][columns]*difference
                response[order, channel, 1] += weight*link[7][columns]*difference
    return response


def complex_tdi2(
    times: np.ndarray,
    latitude: float,
    longitude: float,
    position_splines: Sequence,
    polarizations: Callable[[np.ndarray], tuple[np.ndarray, np.ndarray]],
    *,
    chunk_size: int = 256,
) -> np.ndarray:
    """Evaluate fast X2/Y2/Z2 with directed, first-order delay chains.

    Spacecraft positions in ``position_splines`` are in light-seconds and
    ordered as spacecraft-major Cartesian coordinates.  No C dependency is
    required.  The one-link light time uses the positive quadratic root.
    """
    times = np.asarray(times, dtype=np.float64)
    if times.ndim != 1 or not len(times) or np.any(np.diff(times) <= 0.0):
        raise ValueError("TDI-2 output times must be strictly increasing")
    if len(position_splines) != 9:
        raise ValueError("expected nine spacecraft position splines")
    if chunk_size < 1:
        raise ValueError("chunk_size must be positive")
    propagation, eplus, ecross = _basis(latitude, longitude)
    output = None
    for start in range(0, len(times), chunk_size):
        t = times[start:start+chunk_size]
        links = _reference_links(t, position_splines, propagation, eplus, ecross)
        zeros = np.zeros_like(t)
        ones = np.ones_like(t)
        delay_nodes = {(ch, ()): (zeros, ones) for ch in range(3)}
        term_times = []
        term_coeff = []
        term_channel = []
        for channel in range(3):
            for receiver, emitter, sign, word in _X_TERMS:
                prefix: tuple[int, ...] = ()
                for label in word:
                    next_prefix = prefix+(label,)
                    key = (channel, next_prefix)
                    if key not in delay_nodes:
                        lag, jacobian = delay_nodes[channel, prefix]
                        directed = (((label//10-1+channel) % 3),
                                    ((label%10-1+channel) % 3))
                        light_time, dot_light_time = links[directed][:2]
                        next_lag = lag+light_time-dot_light_time*lag
                        delay_nodes[key] = (next_lag, jacobian*(1.0-dot_light_time))
                    prefix = next_prefix
                lag, jacobian = delay_nodes[channel, prefix]
                directed = (((receiver-1+channel) % 3),
                            ((emitter-1+channel) % 3))
                _, _, ur, ue, dur, due, plus, cross = links[directed]
                term_times.extend((ur-dur*lag, ue-due*lag))
                term_coeff.extend(((sign*jacobian*plus, sign*jacobian*cross),
                                   (-sign*jacobian*plus, -sign*jacobian*cross)))
                term_channel.extend((channel, channel))
        flat_time = np.concatenate(term_times)
        order = np.argsort(flat_time, kind="stable")
        hp_sorted, hc_sorted = polarizations(flat_time[order])
        hp_sorted = np.atleast_2d(np.asarray(hp_sorted, dtype=np.complex128))
        hc_sorted = np.atleast_2d(np.asarray(hc_sorted, dtype=np.complex128))
        if hp_sorted.shape != hc_sorted.shape or hp_sorted.shape[1] != len(flat_time):
            raise ValueError("polarization callback returned inconsistent arrays")
        if output is None:
            output = np.zeros((3, hp_sorted.shape[0], len(times)), dtype=np.complex128)
        elif output.shape[1] != hp_sorted.shape[0]:
            raise ValueError("polarization callback changed its family count")
        hp = np.empty_like(hp_sorted)
        hc = np.empty_like(hc_sorted)
        hp[:, order] = hp_sorted
        hc[:, order] = hc_sorted
        for index, (plus, cross) in enumerate(term_coeff):
            sl = slice(index*len(t), (index+1)*len(t))
            output[term_channel[index], :, start:start+len(t)] += (
                hp[:, sl]*plus+hc[:, sl]*cross
            )
    return output

File: Fast_LISA_Waveforms/c/galactic_binary/tdi2_response.c
/* Copyright (C) 2026 Neil Cornish.
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
#include "tdi2_response.h"

#include <math.h>
#include <stddef.h>

typedef struct {
    int receiver, emitter, sign, count;
    int delays[7];
} Term;

typedef struct {
    double light_time, derivative;
    double receiver_time, emitter_time;
    double receiver_rate, emitter_rate;
    double plus, cross;
} Link;

static const Term x_terms[16] = {
    {1,3, 1,0,{0}}, {3,1, 1,1,{13}},
    {1,2, 1,2,{13,31}}, {2,1, 1,3,{13,31,12}},
    {1,2,-1,0,{0}}, {2,1,-1,1,{12}},
    {1,3,-1,2,{12,21}}, {3,1,-1,3,{12,21,13}},
    {1,2, 1,4,{13,31,12,21}},
    {2,1, 1,5,{13,31,12,21,12}},
    {1,3, 1,6,{13,31,12,21,12,21}},
    {3,1, 1,7,{13,31,12,21,12,21,13}},
    {1,3,-1,4,{12,21,13,31}},
    {3,1,-1,5,{12,21,13,31,13}},
    {1,2,-1,6,{12,21,13,31,13,31}},
    {2,1,-1,7,{12,21,13,31,13,31,12}}
};

static int position(double time, int spacecraft, gsl_interp_accel **acc,
                    gsl_spline **spline, double value[3], double velocity[3])
{
    int coordinate;
    if(!isfinite(time) || time < spline[0]->x[0] ||
       time > spline[0]->x[spline[0]->size-1]) return 1;
    for(coordinate=0; coordinate<3; coordinate++)
    {
        int index = 3*spacecraft+coordinate;
        value[coordinate] = gsl_spline_eval(spline[index], time, acc[index]);
        if(velocity != NULL)
            velocity[coordinate] = gsl_spline_eval_deriv(spline[index], time,
                                                          acc[index]);
    }
    return 0;
}

static int directed_link(double time, int receiver, int emitter,
                         gsl_interp_accel **acc, gsl_spline **spline,
                         const double propagation[3],
                         const double eplus[3][3], const double ecross[3][3],
                         Link *link)
{
    double xr[3], xe[3], vr[3], ve[3], ve0[3], arm[3];
    double r2=0.0, rv=0.0, v2=0.0, length, projection=0.0;
    double pp=0.0, pc=0.0, numerator=0.0, denominator=1.0;
    int i, j;
    if(position(time, receiver, acc, spline, xr, vr) ||
       position(time, emitter, acc, spline, xe, ve0)) return 1;
    for(i=0; i<3; i++)
    {
        double radius = xr[i]-xe[i];
        r2 += radius*radius;
        rv += radius*ve0[i];
        v2 += ve0[i]*ve0[i];
    }
    if(r2 <= 0.0 || v2 >= 1.0) return 2;
    link->light_time = (rv+sqrt(rv*rv+(1.0-v2)*r2))/(1.0-v2);
    if(position(time-link->light_time, emitter, acc, spline, xe, ve)) return 3;
    length = 0.0;
    for(i=0; i<3; i++) length += (xr[i]-xe[i])*(xr[i]-xe[i]);
    length = sqrt(length);
    if(length <= 0.0) return 4;
    link->receiver_time = time;
    link->emitter_time = time-link->light_time;
    link->receiver_rate = 1.0;
    for(i=0; i<3; i++)
    {
        arm[i] = (xr[i]-xe[i])/length;
        projection += propagation[i]*arm[i];
        link->receiver_time -= propagation[i]*xr[i];
        link->emitter_time -= propagation[i]*xe[i];
        link->receiver_rate -= propagation[i]*vr[i];
        numerator += arm[i]*(vr[i]-ve[i]);
        denominator -= arm[i]*ve[i];
    }
    if(fabs(denominator) < 1.0e-12 || fabs(1.0-projection) < 1.0e-12)
        return 5;
    link->derivative = numerator/denominator;
    link->emitter_rate = 1.0-link->derivative;
    for(i=0; i<3; i++) link->emitter_rate -=
        (1.0-link->derivative)*propagation[i]*ve[i];
    for(i=0; i<3; i++) for(j=0; j<3; j++)
    {
        pp += arm[i]*arm[j]*eplus[i][j];
        pc += arm[i]*arm[j]*ecross[i][j];
    }
    link->plus = -0.5*pp/(1.0-projection);
    link->cross = -0.5*pc/(1.0-projection);
    return 0;
}

int gb_tdi2_sample(double output_time, double sky_costheta, double sky_longitude,
                   gsl_interp_accel **position_acc, gsl_spline **position_spline,
                   GBTDIPolarization polarization, void *userdata,
                   double complex output[3])
{
    Link links[3][3];
    double u[3], v[3], propagation[3], eplus[3][3], ecross[3][3];
    double sinth, cosph, sinph;
    int i, j, channel, term_index;
    if(position_acc == NULL || position_spline == NULL ||
       polarization == NULL || output == NULL ||
       fabs(sky_costheta) > 1.0) return 1;
    sinth = sqrt(fmax(0.0, 1.0-sky_costheta*sky_costheta));
    cosph = cos(sky_longitude);
    sinph = sin(sky_longitude);
    u[0] = -sky_costheta*cosph; u[1] = -sky_costheta*sinph; u[2] = sinth;
    v[0] = sinph; v[1] = -cosph; v[2] = 0.0;
    propagation[0] = -sinth*cosph;
    propagation[1] = -sinth*sinph;
    propagation[2] = -sky_costheta;
    for(i=0; i<3; i++) for(j=0; j<3; j++)
    {
        eplus[i][j] = v[i]*v[j]-u[i]*u[j];
        ecross[i][j] = u[i]*v[j]+v[i]*u[j];
    }
    for(i=0; i<3; i++) for(j=0; j<3; j++) if(i != j)
        if(directed_link(output_time, i, j, position_acc, position_spline,
                         propagation, eplus, ecross, &links[i][j])) return 2;
    for(channel=0; channel<3; channel++)
    {
        output[channel] = 0.0;
        for(term_index=0; term_index<16; term_index++)
        {
            const Term *term = &x_terms[term_index];
            int receiver = (term->receiver-1+channel)%3;
            int emitter = (term->emitter-1+channel)%3;
            const Link *link = &links[receiver][emitter];
            double lag=0.0, jacobian=1.0;
            double complex hpr, hcr, hpe, hce;
            for(j=0; j<term->count; j++)
            {
                int label = term->delays[j];
                int dr = (label/10-1+channel)%3;
                int de = (label%10-1+channel)%3;
                const Link *delay = &links[dr][de];
                lag += delay->light_time-delay->derivative*lag;
                jacobian *= 1.0-delay->derivative;
            }
            if(polarization(userdata,
                            link->receiver_time-link->receiver_rate*lag,
                            &hpr, &hcr) ||
               polarization(userdata,
                            link->emitter_time-link->emitter_rate*lag,
                            &hpe, &hce)) return 3;
            output[channel] += term->sign*jacobian*
                (link->plus*(hpr-hpe)+link->cross*(hcr-hce));
        }
    }
    return 0;
}
